larreco/QuadVtx/QuadVtx_module.cc

```diff
@@ -0,0 +1,522 @@
+              std::vector<std::vector<Pt2D>>& pts,
+              std::vector<TVector3>& dirs,
+              const geo::GeometryCore* geom,
+              const detinfo::DetectorProperties* detprop)
+{
+  pts.resize(3); // 3 views
+
+  TVector3 dirZ(0, 0, 1);
+  TVector3 dirU, dirV;
+
+  for(const recob::Hit& hit: hits){
+    const geo::WireID wire = hit.WireID();
+
+    const double xpos = detprop->ConvertTicksToX(hit.PeakTime(), wire);
+
+    const TVector3 r0 = geom->WireEndPoints(wire).start();
+    const TVector3 r1 = geom->WireEndPoints(wire).end();
+
+    const double energy = hit.Integral();
+
+    if(geom->View(hit.Channel()) == geo::kZ){
+      pts[0].emplace_back(xpos, r0.z(), 0, energy);
+      continue;
+    }
+
+    // Compute the direction perpendicular to the wires
+    TVector3 perp = (r1-r0).Unit();
+    perp = TVector3(0, -perp.z(), perp.y());
+    // We want to ultimately have a positive z component in "perp"
+    if(perp.z() < 0) perp *= -1;
+
+    // TODO check we never get a 4th view a-la the bug we had in the 3D version
+
+    // The "U" direction is the first one we see
+    if(dirU.Mag2() == 0){
+      dirU = perp;
+    }
+    else if(dirV.Mag2() == 0 && fabs(dirU.Dot(perp)) < 0.99){
+      // If we still need a "V" and this direction differs from "U"
+      dirV = perp;
+    }
+
+    // Hits belong to whichever view their perpendicular vector aligns with
+    if(fabs(dirU.Dot(perp)) > 0.99){
+      pts[1].emplace_back(xpos, r0.Dot(dirU), 1, energy);
+    }
+    else{
+      pts[2].emplace_back(xpos, r0.Dot(dirV), 2, energy);
+    }
+  } // end for hits
+
+  dirs = {dirZ, dirU, dirV};
+
+  // In case we need to sub-sample they should be shuffled
+  for(int view = 0; view < 3; ++view){
+    std::random_shuffle(pts[view].begin(), pts[view].end());
+  }
+}
+
+// ---------------------------------------------------------------------------
+bool QuadVtx::FindVtx(const std::vector<recob::Hit>& hits,
+                       TVector3& vtx,
+                       art::TFileDirectory* evt_dir) const
+{
+  if(hits.empty()) return false;
+
+  std::vector<std::vector<Pt2D>> pts;
+  std::vector<TVector3> dirs;
+
+  GetPts2D(hits, pts, dirs, geom, detprop);
+
+  std::vector<art::TFileDirectory> view_dirs;
+  if(evt_dir){
+    for(int view = 0; view < 3; ++view){
+      view_dirs.push_back(evt_dir->mkdir(TString::Format("view%d", view).Data()));
+      TGraph* gpts = view_dirs.back().makeAndRegister<TGraph>("hits", "");
+      for(const Pt2D& p: pts[view]) gpts->SetPoint(gpts->GetN(), p.z, p.x);
+    }
+  }
+
+  float minx = +1e9;
+  float maxx = -1e9;
+  float minz[3] = {+1e9, +1e9, +1e9};
+  float maxz[3] = {-1e9, -1e9, -1e9};
+  for(int view = 0; view < 3; ++view){
+    for(const Pt2D& p: pts[view]){
+      minx = std::min(minx, float(p.x));
+      maxx = std::max(maxx, float(p.x));
+      minz[view] = std::min(minz[view], float(p.z));
+      maxz[view] = std::max(maxz[view], float(p.z));
```

Comment:
Recommend either making the `min/max` variables of type `double`, or changing `Pt2D::x` and `Pt2D::z` to be of type `float`.

Reply:
Done